Add document approvals storage and API endpoints.

phgross opened the pull request:
This PR introduces document approvals. The PR adds storage, ApprovalList adapter to manage approvals and API Endpoint additions: The @version endpoints includes now also approvals and the GET Request provides an `approval` expansion.

The feature is currently not used anywhere, but only in the tests. The integration will follow in later stories.
For [CA-2266]

## Checklist

_Everything has to be done/checked. Checked but not present means the author deemed it unnecessary._

- [x] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)

_Only applicable should be left and checked._

- API change:
  - [x] Documentation is updated
  - [ ] API Changelog entry - _I did not add a changelog entry yet, so we can modify the endpoint in later PRs if necessary._


[CA-2266]: https://4teamwork.atlassian.net/browse/CA-2266

njohner commented on opengever/document/approvals.py:
```diff
@@ -62,6 +62,22 @@ def _initialize_storage(self, create_if_missing=False):
+        the given version."""
+
+        if not self._storage:
+            return []
```
no need to return an empty list... just `return`

njohner commented on opengever/document/approvals.py:
```diff
@@ -62,6 +62,22 @@ def _initialize_storage(self, create_if_missing=False):
+        IAnnotations(self.context)[self.ANNOTATIONS_KEY] = except_approvals
+
+    def remove_all(self):
+        IAnnotations(self.context)[self.ANNOTATIONS_KEY] = PersistentList()
```
Why not `IAnnotations(self.context).pop(self.ANNOTATIONS_KEY)`
And dont' you have to check whether the storage exists?

njohner commented on opengever/document/subscribers.py:
```diff
@@ -32,3 +35,17 @@ def set_copyname(doc, event):
+    should be copied."""
+
+    try:
+        current_version = Versioner(event.original).get_current_version_id()
```
What happens for a document with no initial version. Can that document have been approved? If so I guess you should use `missing_as_zero=True` here and probably be careful at other places

phgross replied:
You're right, thx.

njohner commented on opengever/document/tests/test_approvals.py:
```diff
@@ -58,3 +60,32 @@ def test_current_datetime_is_default_approval_time(self):
+            [{'task_uid': IUUID(self.task),
+              'approver': self.regular_user.id,
+              'approved': datetime(2021, 7, 2),
+              'version_id': 2},
```
Do you need to change the `version_id` after copying? I guess the copied document has version_id 0 no?

phgross replied:
Indeed, thx.

njohner reviewed (commented):
LGTM just a few comments. Main point is what happens with documents without initial versions?

phgross commented:
@njohner thx for your review! I updated my PR.

For documents without an initial version, the approvals are still handled for version 0. i think this is correct. Because a document without a version is in fact in version 0 and it is actually only a technical detail that this does not really exist yet. Do you see other problems/issues? 🤔

phgross commented on opengever/api/tests/test_document.py:
```diff
@@ -187,15 +187,28 @@ def test_approvals_expansion(self, browser):
+                  u'@id': u'http://nohost/plone/ordnungssystem/fuhrung/vertrage-und-vereinbarungen/dossier-1/task-1/task-2',
+                  u'@type': u'opengever.task.task',
+                  u'description': u'',
+                  u'is_leafnode': None,
```
Will fix that in a separate PR.

njohner commented on opengever/document/approvals.py:
```diff
@@ -0,0 +1,131 @@
+        for the current version.
+        """
+        if not current_version:
+            self.storage.remove_all()
```
return

njohner commented on opengever/document/approvals.py:
```diff
@@ -0,0 +1,131 @@
+        self.storage.add(approved, approver, IUUID(task), version_id)
+
+    def get(self):
+        return [Approval(**item) for item in self.storage.list()]
```
`return [approval for approval in self]` ? Maybe even `return list(self)` would work😅

njohner reviewed (approved):
LGTM thanks! There is one more concern with permissions IMO.

For documents without initial version I cannot judge whether everything works as expected, but I would expect:
- Approving a document without initial version should store `version 0` in the approval list
- When copying a document without initial version but an approval, that approval should be copied with
- Serialization of a document without initial version but an approval should return that approval as `version 0`

phgross commented on opengever/api/serializer.py:
```diff
@@ -178,6 +180,20 @@ def oguid_converter(value):
+@adapter(Approval)
+@implementer(IJsonCompatible)
+def approval_converter(approval):
+    task = getMultiAdapter((approval.get_task_brain(), getRequest()),
```
I updated my PR, the serializer returns None on the `task` key, if the task is not accessible, see last commit.

phgross commented:
@njohner I updated my PR. Added a test for documents without a version and also handle the "permission-problem" if the task is not accessible in the serializer.

njohner reviewed (approved):
LGTM thanks!

njohner commented:
I'm merging this one, as philippe is on holidays...
